### PR TITLE
STM32: AnalogIO

### PR DESCRIPTION
**Pull request by hierophect**
This PR will add the AnalogIO module to STM32. Confirmed functional on the STM32F411, but depends on pending changes to bitpacking style and USB via the general pin remap in #2126. Also looking for style/readability input on the use of the LL libraries vs HAL.

### CODE REVIEW

**Comment by tannewt** [comment]
In general, I dislike the "fill out a large struct and pass it in style" that HALs typically have because a lot of code is unneeded. However, if the HAL will make it easier to use code across stm families then it might be worth it.

I'm not sure if the low level code is best because it can cause consecutive writes to the same register rather than setting things all at once. Is it possible to write to a register directly via a struct? Does the LL API abstract us across families still?

**Comment by tannewt** [comment]
HAL is tine with me for now if you think it's best. We can always optimize later.

**Comment by hierophect** [comment]
@ladyada this should now be ready to test on a F412, had some usb issues.

**Comment by hierophect** [comment]
New binary with reset fix
[firmware.bin.zip](https://github.com/adafruit/circuitpython/files/3610217/firmware.bin.zip)

**Comment by hierophect** [comment]
Currently experiencing a sort of timed soft-reboot after 30 seconds on the first save to the F412. Also occurs on the peripherals branch and possibly on earlier branches. Only happens once.

**Comment by ladyada** [comment]
```# CircuitPython AnalogIn Demo
import time
import board
from analogio import AnalogIn
 
analogs = []
for pin in (board.A0, board.A1, board.A2, board.A3, board.A4, board.A5):
    analogs.append(AnalogIn(pin))
 
def get_voltage(pin):
    return (pin.value * 3.3) / 65536
 
while True:
    print("(", end='')
    for a in analogs:
        print("%0.1f,\t" % get_voltage(a), end='')
    print(")")
    time.sleep(0.1)```

trying the firmware.bin you posted, only board.A1 works, and the others 'follow' A1. that said, A1 does work! so its something in the bin tables

**Comment by hierophect** [comment]
I’ll check that out tomorrow morning, thanks for the heads up. I was using
the direct PIN numbers for the Analog channels so maybe something is messed
up in the aliases.

**Comment by ladyada** [comment]
kk also digitalio stopped toggling output pins, like it doesnt complain or anything but when i run my LED blinky demo no LEDs light up

**Comment by hierophect** [comment]
@ladyada your analog testing sketch gives me flash corruption so extreme it actually disables the debug program on my mac and I have to erase on the official ST utility on windows to enable the chip for use again 😲

Hopefully won't be too long to a fix but without a debugger it could be tricky.

**Comment by ladyada** [comment]
hmm thats interesting, it didnt get me flash corruption on win10 but it definitely has very slow flash writes

**Comment by tannewt** [comment]
Doesn't an ST Link work as a debugger? What am I missing?

**Comment by hierophect** [comment]
@tannewt things have been a bit inconsistent, but my texane/stlink program which launches the gdb server does not recognize the flash of the F412 as valid, and cannot load binaries. The F411 was having the same issue, but I was actually able to fix it by erasing it with the windows utility (no luck doing the same for the F412). Latest Analogio code will irrecoverably crash mu (computer must be restarted), and corrupt the flash, making both boards un-debuggable as well as unwritable. Can't get openocd working either, and it's not even printing complete errors... 

Anyway I'm deep in bug hell. I'll update once I've climbed out.

**Comment by tannewt** [comment]
Ok, ping me tomorrow morning if you haven't gotten out by the end of today.

**Comment by hierophect** [comment]
I think I should be. While I reset-protected pins that seemed obviously critical like crystal pins and debugging, I think I must have missed something else that was system critical. My suspicion is that GPIO resets on some pins might even persist past a flash erase. I'm going to invert the never reset process to exclude all pins that aren't directly broken out by the board, and see if that helps.

**Comment by hierophect** [comment]
The latest commit should fix all issues, hopefully. @ladyada, I've attached a binary if you'd like to try it out again, or alternatively you could try grabbing the binary out of the CI as Scott's been promoting. 
[firmware.bin.zip](https://github.com/adafruit/circuitpython/files/3623590/firmware.bin.zip)

**Comment by ladyada** [comment]
ok will try tonite!

**Comment by ladyada** [comment]
tested! this has fixed the digitalio AND analogin stuff - can now digitalio and have LEDs toggle, and analog in works independantly on A0-A5

nice job! want me to merge or you have things you still want to do with it?

**Comment by hierophect** [comment]
I'm fine with a merge if you are! It'll get one more thing off my list :)

**Comment by ladyada** [comment]
ok i kicked github CI - might hve to do it again, its being hang-y once it passes ill merge

**Comment by ladyada** [comment]
@hierophect can you do like a fake push with a whitespace change? i cant seem to restart CI

**Comment by tannewt** [comment]
Note for next time, if you see the "not a tree warning" the best thing to do is merge in the latest master and push it. It only happens if something else has been merged into master since the *start* of the initial CI run. This is why it happens even on the first run if master changes.

File: ports/stm32f4/common-hal/analogio/AnalogIn.c [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    //     LL_ADC_SetChannelSamplingTime(ADCx, self->pin->adc_channel, LL_ADC_SAMPLINGTIME_56CYCLES);
+    //     LL_ADC_EnableIT_OVR(ADCx);
+    // }
+    // LL_ADC_Enable(ADCx);
+    // uint16_t uhADCxConvertedData = (__LL_ADC_DIGITAL_SCALE(LL_ADC_RESOLUTION_12B) + 1);
+    // LL_ADC_REG_StartConversionSWStart(ADCx);
+    // while (LL_ADC_IsActiveFlag_EOCS(ADCx) == 0) {}
+    // /* Retrieve ADC conversion data */
+    // /* (data scale corresponds to ADC resolution: 12 bits) */
+    // uhADCxConvertedData = LL_ADC_REG_ReadConversionData12(ADCx);
```
Comment (tannewt) [inline ports/stm32f4/common-hal/analogio/AnalogIn.c]: Please remove the above because it is untested.

**Comment by tannewt** [comment]
Feel free to wait until the CI finishes to know if there are any other last fixes needed.

**Comment by ladyada** [comment]
@hierophect please fix above deletion request - WE ARE SO CLOSE :)

**Pull request merged**